[WK2] Provide generic callReply(), cancelReply() functions instead of per-message variants

zdobersek opened the pull request:
#### 069295b2bd5be1a999e9187001c361c602415b91
<pre>
[WK2] Provide generic callReply(), cancelReply() functions instead of per-message variants
<a href="https://bugs.webkit.org/show_bug.cgi?id=248132">https://bugs.webkit.org/show_bug.cgi?id=248132</a>

Reviewed by Darin Adler.

For each asynchronous IPC message, the callReply() and cancelReply()
static functions are generated. The first decodes each specified reply
argument before invoking the completion handler, bailing if the decoding
failed at any point. The second cancels the reply by invoking the
completion handler with an AsyncReplyError value for every reply
parameter.

These distinct functions can be replaced by single generic variants that
work with the ReplyArguments tuple that&apos;s defined for each message that
expects a reply. Connection::callReply() decodes a tuple of reply
arguments before invoking the completion handler, bailing if decoding
failed. Connection::cancelReply() also invokes the completion handler,
but providing AsyncReplyError values for each reply parameter.

Different header inclusions are added to avoid incomplete types. A solid
chunk of code generation is removed from messages.py, and unit tests are
updated to reflect the reduction of generated code.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::makeAsyncReplyHandler):
(IPC::Connection::callReply):
(IPC::Connection::cancelReply):
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type):
(message_to_struct_declaration):
(generate_message_handler):
(reply_parameter_type): Deleted.
(move_type): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
(Messages::TestWithCVPixelBuffer::ReceiveCVPixelBuffer::callReply): Deleted.
(Messages::TestWithCVPixelBuffer::ReceiveCVPixelBuffer::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
(Messages::TestWithImageData::ReceiveImageData::callReply): Deleted.
(Messages::TestWithImageData::ReceiveImageData::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(Messages::TestWithLegacyReceiver::CreatePlugin::callReply): Deleted.
(Messages::TestWithLegacyReceiver::CreatePlugin::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::RunJavaScriptAlert::callReply): Deleted.
(Messages::TestWithLegacyReceiver::RunJavaScriptAlert::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::GetPlugins::callReply): Deleted.
(Messages::TestWithLegacyReceiver::GetPlugins::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::InterpretKeyEvent::callReply): Deleted.
(Messages::TestWithLegacyReceiver::InterpretKeyEvent::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
(Messages::TestWithSemaphore::ReceiveSemaphore::callReply): Deleted.
(Messages::TestWithSemaphore::ReceiveSemaphore::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessageReceiver.cpp:
(Messages::TestWithSuperclass::TestAsyncMessage::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessage::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithNoArguments::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithNoArguments::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithMultipleArguments::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithMultipleArguments::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithConnection::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithConnection::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(Messages::TestWithoutAttributes::CreatePlugin::callReply): Deleted.
(Messages::TestWithoutAttributes::CreatePlugin::cancelReply): Deleted.
(Messages::TestWithoutAttributes::RunJavaScriptAlert::callReply): Deleted.
(Messages::TestWithoutAttributes::RunJavaScriptAlert::cancelReply): Deleted.
(Messages::TestWithoutAttributes::GetPlugins::callReply): Deleted.
(Messages::TestWithoutAttributes::GetPlugins::cancelReply): Deleted.
(Messages::TestWithoutAttributes::InterpretKeyEvent::callReply): Deleted.
(Messages::TestWithoutAttributes::InterpretKeyEvent::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/RemoteWebLockRegistry.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::MockTestMessageWithAsyncReply1::callReply): Deleted.
(TestWebKitAPI::MockTestMessageWithAsyncReply1::cancelReply): Deleted.

Canonical link: <a href="https://commits.webkit.org/256936@main">https://commits.webkit.org/256936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5683916dc4050020666aabfbf22f875d85727eb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97291 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106810 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167076 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6851 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35290 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89688 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103491 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102954 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83920 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32134 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/100871 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86998 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75060 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/571 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20289 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/555 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21748 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4781 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5356 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1799 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41077 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -372,22 +372,10 @@ class Connection : public ThreadSafeRefCounted<Connection, WTF::DestructionThrea
-            nextAsyncReplyHandlerID()
-        };
-    }
+    template<typename T, typename C> static AsyncReplyHandler makeAsyncReplyHandler(C&& completionHandler, ThreadLikeAssertion callThread = CompletionHandlerCallThread::AnyThread);
```
Can these be `private`?

zdobersek replied:
`makeAsyncReplyHandler()` has to be public, it's also used from  MessageSender and AuxiliaryProcessProxy classes. The new ones can be private, so I'll make them.

darinadler commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -674,6 +662,49 @@ inline std::unique_ptr<Decoder> Connection::waitForMessageForTesting(MessageName
+        completionHandler();
+    } else {
+        auto replyArguments = decoder.decode<typename T::ReplyArguments>();
+        if (!!replyArguments) {
```
Is the `!!` required here? We Rey to avoid that in WebKit code. I suggest:
```
if (auto arguments = decoder.decode<typename T::ReplyArguments>()) {
```

zdobersek replied:
Adopted.

darinadler reviewed (approved):
Looks great

zdobersek committed:
[WK2] Provide generic callReply(), cancelReply() functions instead of per-message variants
https://bugs.webkit.org/show_bug.cgi?id=248132

Reviewed by Darin Adler.

For each asynchronous IPC message, the callReply() and cancelReply()
static functions are generated. The first decodes each specified reply
argument before invoking the completion handler, bailing if the decoding
failed at any point. The second cancels the reply by invoking the
completion handler with an AsyncReplyError value for every reply
parameter.

These distinct functions can be replaced by single generic variants that
work with the ReplyArguments tuple that's defined for each message that
expects a reply. Connection::callReply() decodes a tuple of reply
arguments before invoking the completion handler, bailing if decoding
failed. Connection::cancelReply() also invokes the completion handler,
but providing AsyncReplyError values for each reply parameter.

Different header inclusions are added to avoid incomplete types. A solid
chunk of code generation is removed from messages.py, and unit tests are
updated to reflect the reduction of generated code.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::Connection::makeAsyncReplyHandler):
(IPC::Connection::callReply):
(IPC::Connection::cancelReply):
* Source/WebKit/Scripts/webkit/messages.py:
(function_parameter_type):
(message_to_struct_declaration):
(generate_message_handler):
(reply_parameter_type): Deleted.
(move_type): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessageReceiver.cpp:
(Messages::TestWithCVPixelBuffer::ReceiveCVPixelBuffer::callReply): Deleted.
(Messages::TestWithCVPixelBuffer::ReceiveCVPixelBuffer::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithCVPixelBufferMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessageReceiver.cpp:
(Messages::TestWithImageData::ReceiveImageData::callReply): Deleted.
(Messages::TestWithImageData::ReceiveImageData::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithImageDataMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessageReceiver.cpp:
(Messages::TestWithLegacyReceiver::CreatePlugin::callReply): Deleted.
(Messages::TestWithLegacyReceiver::CreatePlugin::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::RunJavaScriptAlert::callReply): Deleted.
(Messages::TestWithLegacyReceiver::RunJavaScriptAlert::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::GetPlugins::callReply): Deleted.
(Messages::TestWithLegacyReceiver::GetPlugins::cancelReply): Deleted.
(Messages::TestWithLegacyReceiver::InterpretKeyEvent::callReply): Deleted.
(Messages::TestWithLegacyReceiver::InterpretKeyEvent::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithLegacyReceiverMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessageReceiver.cpp:
(Messages::TestWithSemaphore::ReceiveSemaphore::callReply): Deleted.
(Messages::TestWithSemaphore::ReceiveSemaphore::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithSemaphoreMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithStreamMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessageReceiver.cpp:
(Messages::TestWithSuperclass::TestAsyncMessage::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessage::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithNoArguments::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithNoArguments::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithMultipleArguments::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithMultipleArguments::cancelReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithConnection::callReply): Deleted.
(Messages::TestWithSuperclass::TestAsyncMessageWithConnection::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithSuperclassMessages.h:
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessageReceiver.cpp:
(Messages::TestWithoutAttributes::CreatePlugin::callReply): Deleted.
(Messages::TestWithoutAttributes::CreatePlugin::cancelReply): Deleted.
(Messages::TestWithoutAttributes::RunJavaScriptAlert::callReply): Deleted.
(Messages::TestWithoutAttributes::RunJavaScriptAlert::cancelReply): Deleted.
(Messages::TestWithoutAttributes::GetPlugins::callReply): Deleted.
(Messages::TestWithoutAttributes::GetPlugins::cancelReply): Deleted.
(Messages::TestWithoutAttributes::InterpretKeyEvent::callReply): Deleted.
(Messages::TestWithoutAttributes::InterpretKeyEvent::cancelReply): Deleted.
* Source/WebKit/Scripts/webkit/tests/TestWithoutAttributesMessages.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/RemoteWebLockRegistry.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
(TestWebKitAPI::MockTestMessageWithAsyncReply1::callReply): Deleted.
(TestWebKitAPI::MockTestMessageWithAsyncReply1::cancelReply): Deleted.

Canonical link: https://commits.webkit.org/256936@main
